### PR TITLE
[node-red-contrib-binpacking]  adapt bin packing algorithm for no-stackable packages

### PR DESCRIPTION
**Pull request by mrazafindrazaka**
Adapted algorithm for no-stackable packages and test with jest + supertest

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -43,7 +45,7 @@
   "scripts": {
     "lint": "eslint --ext .js index.js",
     "lint-fix": "eslint --fix --ext .js index.js",
-    "test": "npm run lint",
```
Comment (acacopardo) [inline package.json:46]: test commande should run all test lint + jest.
Can you di a specific commande for jest test and modify test commande to run jest and test npm commande please ?

**Comment by acacopardo** [comment]
Can you remove package.json.lock please ?
for a library is not recommended.

**Review by gbo-redspher — APPROVED** [review]
Code-wise it seems fine for me, for the business logic part I'd rather wait for @fanshan or @glickel review

File: index.js [hunk 2]
```diff
@@ -36,11 +49,21 @@ module.exports = function (RED) {
       if (msg.packages && msg.packages.length > 0) {
         for (const pkg in msg.packages) {
           if (!msg.packages[pkg].quantity) msg.packages[pkg].quantity = 1
+          if (!Object.prototype.hasOwnProperty.call(msg.packages[pkg], 'stackable')) {
```
Comment (glickel) [inline index.js:52]:
```suggestion
          if (typeof msg.packages[pkg].stackable === 'undefined') {
```

Comment (glickel) [reply index.js:52]: Easier to read

File: index.js [hunk 3]
```diff
@@ -36,11 +49,21 @@ module.exports = function (RED) {
...
           if (msg.packages[pkg].width && msg.packages[pkg].height && msg.packages[pkg].length && msg.packages[pkg].weight) {
             for (let q = 0; q < msg.packages[pkg].quantity; q++) {
               if (!msg.packages[pkg].name) msg.packages[pkg].name = 'Item'
               if (!msg.packages[pkg].allowedRotation) msg.packages[pkg].allowedRotation = [0, 1, 2, 3, 4, 5]
-              packages.push(new Item(msg.packages[pkg].name + ' ' + q, msg.packages[pkg].width, msg.packages[pkg].height, msg.packages[pkg].length, msg.packages[pkg].weight, msg.packages[pkg].allowedRotation))
+              packages.push(
+                {
+                  item: new Item(msg.packages[pkg].name + ' ' + q, msg.packages[pkg].width,
+                    msg.packages[pkg].height, msg.packages[pkg].length,
+                    msg.packages[pkg].weight, msg.packages[pkg].allowedRotation),
```
Comment (glickel) [inline index.js:63]:
```suggestion
                  item: new Item(
                    msg.packages[pkg].name + ' ' + q, msg.packages[pkg].width,
                    msg.packages[pkg].height, msg.packages[pkg].length,
                    msg.packages[pkg].weight, msg.packages[pkg].allowedRotation
                  ),
```

Easier to read

**Review by glickel — CHANGES_REQUESTED** [review]

**Review by glickel — APPROVED** [review]

File: package.json [hunk 4]
```diff
@@ -43,7 +45,8 @@
...
+    "jest": "jest",
+    "test": "npm run lint && npm run jest",
     "start": "node-red -s sample/settings.js sample/flow.json -u sample/"
   },
   "version": "1.2.3"
```
Comment (acacopardo) [inline package.json:52]:
```suggestion
  "version": "1.3.0"
```

**Review by acacopardo — APPROVED** [review]

**Pull request merged**